src/ui/api/repository_label.go

```diff
@@ -0,0 +1,248 @@
+		if err != nil {
+			r.HandleInternalServerError(fmt.Sprintf("failed to get label %d: %v", labelID, err))
+			return
+		}
+
+		if label == nil {
+			r.HandleNotFound(fmt.Sprintf("label %d not found", labelID))
+			return
+		}
+		r.label = label
+	}
+}
+
+// GetOfImage returns labels of an image
+func (r *RepositoryLabelAPI) GetOfImage() {
+	r.getLabels(common.ResourceTypeImage, fmt.Sprintf("%s:%s", r.repository.Name, r.tag))
+}
+
+// AddToImage adds the label to an image
+func (r *RepositoryLabelAPI) AddToImage() {
+	rl := &models.ResourceLabel{
+		LabelID:      r.label.ID,
+		ResourceType: common.ResourceTypeImage,
+		ResourceID:   fmt.Sprintf("%s:%s", r.repository.Name, r.tag),
+	}
+	r.addLabel(rl)
+}
+
+// RemoveFromImage removes the label from an image
+func (r *RepositoryLabelAPI) RemoveFromImage() {
+	rl := &models.ResourceLabel{
+		LabelID:      r.label.ID,
+		ResourceType: common.ResourceTypeImage,
+		ResourceID:   fmt.Sprintf("%s:%s", r.repository.Name, r.tag),
+	}
+	r.removeLabel(rl)
+}
+
+// GetOfRepository returns labels of a repository
+func (r *RepositoryLabelAPI) GetOfRepository() {
+	r.getLabels(common.ResourceTypeRepository, strconv.FormatInt(r.repository.RepositoryID, 10))
+}
+
+// AddToRepository adds the label to a repository
+func (r *RepositoryLabelAPI) AddToRepository() {
+	rl := &models.ResourceLabel{
+		LabelID:      r.label.ID,
+		ResourceType: common.ResourceTypeRepository,
+		ResourceID:   strconv.FormatInt(r.repository.RepositoryID, 10),
+	}
+	r.addLabel(rl)
+}
+
+// RemoveFromRepository removes the label from a repository
+func (r *RepositoryLabelAPI) RemoveFromRepository() {
+	rl := &models.ResourceLabel{
+		LabelID:      r.label.ID,
+		ResourceType: common.ResourceTypeRepository,
+		ResourceID:   strconv.FormatInt(r.repository.RepositoryID, 10),
+	}
+	r.removeLabel(rl)
+}
+
+func (r *RepositoryLabelAPI) getLabels(rType, rID string) {
+	labels, err := dao.GetLabelsOfResource(rType, rID)
+	if err != nil {
+		r.HandleInternalServerError(fmt.Sprintf("failed to get labels of resource %s %s: %v",
+			rType, rID, err))
+		return
+	}
+	r.Data["json"] = labels
+	r.ServeJSON()
+}
+
+func (r *RepositoryLabelAPI) addLabel(rl *models.ResourceLabel) {
+	rlabel, err := dao.GetResourceLabel(rl.ResourceType, rl.ResourceID, rl.LabelID)
+	if err != nil {
+		r.HandleInternalServerError(fmt.Sprintf("failed to check the existence of label %d for resource %s %s: %v",
+			rl.LabelID, rl.ResourceType, rl.ResourceID, err))
+		return
+	}
+
+	if rlabel != nil {
+		r.HandleConflict()
+		return
+	}
+	if _, err := dao.AddResourceLabel(rl); err != nil {
+		r.HandleInternalServerError(fmt.Sprintf("failed to add label %d to resource %s %s: %v",
+			rl.LabelID, rl.ResourceType, rl.ResourceID, err))
+		return
+	}
+
+	// return the ID of label and return status code 200 rather than 201 as the label is not created
+	r.Redirect(http.StatusOK, strconv.FormatInt(rl.LabelID, 10))
+}
+
+func (r *RepositoryLabelAPI) removeLabel(rl *models.ResourceLabel) {
```

Comment:
The resource_label should be removed if the label is deleted.

Reply:
Open #4410 to track this, will submit another PR to fix it.